components/function-controller/design/extend_function_spec_proposal.md

```diff
@@ -0,0 +1,392 @@
+
+### Goals
+
+- Add more flexibility to the Serverless API - enable volume mounted secrets and pod annotations.
+- Organise spec attributes belonging to runtime and build-time configuration (?)
+- Provide a rollout plan to implement this functionality in small increments and avoid the need to roll out a new API version
+
+### Non-Goals
+- Define a new API version specification
+
+## Discussion points
+
+### Runtime, build-time separation
+
+Since Function CR is managing two workloads ( Deployment for runtime and Job for build-time ) we need to separate them in the spec in order to make it clear where do the mounts and annotations belong. 
+
+We can either make a clear separation, i.e:
+
+```yaml 
+spec:
+  source:
+  runtimeSpec:
+    # mounts
+    # resources
+    # envs
+    # metadata
+  buildSpec:
+    # resources
+    # envs (?)
+    # metadata (?)
+```
+
+Alternatively we could promote runtime fields to the root (as the belong to kind: Function) and extract only the build-time fields 
+
+```yaml 
+spec:
+  source:
+  # mounts
+  # resources
+  # envs
+  # metadata
+  buildSpec:
+    # resources
+    # envs (?)
+    # metadata (?)
+```
+
+### Mounts - own structure or k8s inherited
+
+Under the hood the secret mount becomes a volume mount in the runtime pod.
+We could :
+ - expose the k8s volume mount spec  in function spec
+ - cover k8s volume mount spec via oor own [specialised](https://servicebinding.io/application-developer/) spec
+
+```yaml 
+apiVersion: serverless.kyma-project.io/v1alpha2
+kind: Function
+spec:
+  volumeMounts:
+  - name: foo
+    mountPath: "/etc/foo"
+    readOnly: true
+  volumes:
+  - name: foo
+    secret:
+      secretName: mysecret
+```
+
+```yaml 
+apiVersion: serverless.kyma-project.io/v1alpha2
+kind: Function
+spec:
+  serviceBindings:
+  - source: my-secret
+  env:
+  - name: SERVICE_BINDING_ROOT 
+    value: /foo
+```
+
+### Metadata for function pod
+
+Define runtime labels and annotations on the root level or under a `metadata` field
+
+```yaml
+apiVersion: serverless.kyma-project.io/v1alpha2
+kind: Function
+metadata:
+  name: my-function
+  namespace: default
+  labels:
+    ...
+  annotations:
+    ...
+spec:
+  metadata: 
+    labels: ...
+    annotations: ...
+```
+OR
+```yaml
+apiVersion: serverless.kyma-project.io/v1alpha2
+kind: Function
+metadata:
+  name: my-function
+  namespace: default
+  labels:
+    ...
+  annotations:
+    ...
+spec:
+  labels: ...
+  annotations: ...
+```
+
+## Samples: 
+
+### Option 1
+
+- simplified mounts serving just for service binding purpose
+- extract `build` object for any build-time specific config
+- runtime labels and annotations on the spec root level
+
+```yaml
+apiVersion: serverless.kyma-project.io/v1alpha2
+kind: Function
+metadata:
+  name: my-function
+  namespace: default
+  labels:
+    app.kubernetes.io/name: my-function
+spec:
+  runtime: nodejs16
+  source:
+    ...
+  serviceBindings: 
```

Comment:
This is an interesting name. I think it makes sense.

Reply:
the final name would have to be `secretBindings` though as `service binding` would clash with https://github.com/SAP/sap-btp-service-operator#service-binding

Reply:
in the end the implicit associations with service bindings (https://github.com/SAP/sap-btp-service-operator#service-binding or https://servicebinding.io/) turn out to be unwanted. 
To keep the API not clean from this domain the proposal was changed to `secretMounts`